gh-128807: Add marking phase for free-threaded cyclic GC

nascheme opened the pull request:
This is conceptually similar to the phase that was added to the non-free-threaded GC.  Start with a set of known root objects, like sysdict and mark all objects reachable from those (revealed by the tp_traverse method) as "alive".  We know anything marked alive cannot be garbage and can be excluded from the regular cyclic GC process.  For most programs, this saves a moderate amount of computation since the marking pass is relatively cheaper per object.

If `gc.freeze()` is used, it's unlikely that this marking phase will be a win since it's expected that the majority of objects will be frozen.  Disable the marking phase if freeze is used.

See gh-126491 for the non-free-threaded version of this technique.

pyperformance [results vs merge base](https://github.com/facebookexperimental/free-threading-benchmarking/blob/main/results/bm-20250113-3.14.0a3%2B-0c173c9-NOGIL/bm-20250113-vultr-x86_64-nascheme-nogil_gc_mark_alive-3.14.0a3%2B-0c173c9-vs-base.svg).  I suspect the slowdown on some benchmarks is not real.  For example, regex_v8 should not be slower.

[Here](https://gist.github.com/nascheme/e18406f977b3d37ef3d9813320e8faa9) are the pyperformance results for a bare-metal AMD Ryzen machine.  It does not show a slowdown on regex_v8, for example.

To better show the expected improvement, I ran a "sphinx build" benchmark, like in gh-124567.  Results are:

|   | old  | new  |
|---|---|---|
| GC collections  | 38  | 38  | 
| long-lived objects  |  586,986  |  585,730  |
| total run time  | 3.24 s | 3.03 s  |
| mark phase time | 0.00 s | 0.16 s |
| total gc time | 0.72 s | 0.57 s |



<!-- gh-issue-number: gh-128807 -->
* Issue: gh-128807
<!-- /gh-issue-number -->

nascheme committed:
wip: gc mark alive for nogil

nascheme committed:
wip: mark more roots

nascheme committed:
wip: mark stack refs as alive

nascheme committed:
wip: add helper gc_maybe_untrack()

Reduces duplicate code.

nascheme committed:
wip: revert changes to gc_should_collect()

These are not strictly needed, simplify PR.

nascheme committed:
wip: fix reversion on mark_heap_visitor()

If the object is already marked as reachable, we shouldn't traverse it
again.

nascheme committed:
wip: code cleanup, small bug fix

More code cleanup (better names, comments, simplify error handling).
Fix bug in that "alive" bit must be checked in mark_alive_stack_push()
to avoid visiting already visited objects.

nascheme committed:
Add NEWS entry.

nascheme committed:
wip: log gc timing to temp file

nascheme committed:
wip: include # of collected in debug log

nascheme committed:
wip: move freeze_used to gc state

nascheme committed:
wip: fix OOM error handling, add ifdef toggles

nascheme committed:
wip: removing timing code

nascheme committed:
wip: untrack tuples in "mark alive" pass

Make sure we still do this optimization.  There is also a unit test that
checks for this.

nascheme committed:
wip: enable stacks and extra roots

nascheme committed:
wip: spelling fixes, minor code cleanup

nascheme committed:
Merge branch 'origin/main' into nogil-gc-mark-alive

colesbury commented on Python/gc_free_threading.c:
```diff
@@ -392,6 +446,94 @@ gc_visit_thread_stacks(PyInterpreterState *interp)
+
+#ifdef GC_MARK_ALIVE_STACKS
+static bool
+gc_visit_stackref_mark_alive(_PyObjectStack *stack, _PyStackRef stackref)
```
I think it's a bit clearer to stick with one return/error code convention, such as `int` with `-1=error` instead of having some functions use `bool`.

nascheme replied:
I'm not a big fan of that -1 = error convention, where only 0 and -1 are the returned values.  However, it's pretty common in the CPython codebase and I agree that it's better to be consistent.  I will change it.

colesbury commented on Python/gc_free_threading.c:
```diff
@@ -1245,6 +1491,19 @@ gc_collect_internal(PyInterpreterState *interp, struct collection_state *state,
     process_delayed_frees(interp, state);
 
+    #ifdef GC_ENABLE_MARK_ALIVE
+    if (!state->gcstate->freeze_active) {
```
I think it's worth including your reasoning from the PR explaining the motivation for this condition as a code comment.

colesbury commented on Python/gc_free_threading.c:
```diff
@@ -679,13 +846,92 @@ scan_heap_visitor(const mi_heap_t *heap, const mi_heap_area_t *area,
+
+    #define STACK_PUSH(op) \
+        if (mark_alive_stack_push(op, &stack) < 0) { \
+            return -1; \
```
On cleanup:

- I'd normally expect the cleanup of `stack` (the `_PyObjectStack_Clear()`) to happen at the same level as its creation (i.e., in this function instead of in `mark_alive_stack_push`)
- I think we may need to clear all the alive bits on error.

colesbury reviewed (commented):
Looks great! A few comments below

nascheme committed:
wip: use -1 for error convention for functions

nascheme committed:
wip: improve error handling for OOM

Use gc_abort_mark_alive() helper in case of OOM.  In addition to
freeing the stack, we need to ensure that no object has the alive bit
set on it.  This also adding missing error handling in the case that
propagate_alive_bits() fails.

nascheme committed:
Add comment about gc.freeze() disabling marking.

colesbury commented on Python/gc_free_threading.c:
```diff
@@ -679,13 +871,97 @@ scan_heap_visitor(const mi_heap_t *heap, const mi_heap_area_t *area,
+    }
+#endif
+#ifdef GC_MARK_ALIVE_STACKS
+    if (gc_visit_thread_stacks_mark_alive(interp, &stack) < 0) {
```
I think that if we are marking the stacks here, then we don't need to visit them in `deduce_unreachable_heap`, but it's not super important.

nascheme replied:
Yeah, I think you're right.  I'll address in a follow up PR.

kumaraditya303 commented on Python/gc_free_threading.c:
```diff
@@ -392,6 +446,119 @@ gc_visit_thread_stacks(PyInterpreterState *interp)
+
+#ifdef GC_ENABLE_MARK_ALIVE
+static int
+mark_alive_stack_push(PyObject *op, _PyObjectStack *stack)
```
I would expect the first arg to be stack here from the name, maybe change the signature to `mark_alive_stack_push(_PyObjectStack *stack, PyObject *op)`

nascheme replied:
The object needs to come first since this function is passed to `tp_traverse`.

colesbury commented on Python/gc_free_threading.c:
```diff
@@ -679,13 +871,97 @@ scan_heap_visitor(const mi_heap_t *heap, const mi_heap_area_t *area,
+
+    #define STACK_PUSH(op) \
+        if (mark_alive_stack_push(op, &stack) < 0) { \
+            gc_abort_mark_alive(interp, state, &stack); \
```
I wonder if it's worth calling `propagate_alive_bits()` earlier and more frequently instead of building up a big marking stack.

In other words, any time we call `mark_alive_stack_push()`, we could call `propagate_alive_bits` immediately.

nascheme replied:
I expect the pre-fetching will change how this works so I'll defer this until I work on that.  Keeping that marking stack small would be good.

colesbury reviewed (approved):
LGTM
